cunumeric/array.py

```diff
@@ -3728,3 +3754,70 @@ def _perform_where(cls, mask, one, two) -> ndarray:
         out = ndarray(shape=out_shape, dtype=common_type, inputs=args)
         out._thunk.where(mask, one, two)
         return out
+
+    @classmethod
+    def _perform_scan(
+        cls, op, src, axis=None, dtype=None, out=None, nan0=False
+    ):
+        if dtype is None:
+            if out is None:
+                if np.issubdtype(src.dtype, np.integer):
+                    # Set dtype to default platform integer
+                    dtype = np.int_
+                else:
+                    dtype = src.dtype
+            else:
+                dtype = out.dtype
+        if (
+            np.issubdtype(src.dtype, np.floating) or np.iscomplexobj(src)
+        ) and np.issubdtype(dtype, np.integer):
+            # Needs changes to convert()
+            raise NotImplementedError(
+                "Integer output types currently not supported for "
+                "floating/complex inputs"
+            )
+        if np.issubdtype(dtype, np.complexfloating) and nan0 is True:
+            # Currently not behaving correctly
+            raise NotImplementedError(
+                "nancumsum and nancumprod currently do not "
+                "supported complex types"
+            )
+        if out is not None:
+            if dtype != out.dtype:
+                # if out array is specified, its type overrules dtype
+                dtype = out.dtype
+            if axis is not None:
+                assert out.shape == src.shape
```

Comment:
In case it is useful: some additional context on asserts vs exceptions. 

Assertions are useful to maintain *internal invariants*. Commonly, when a function passes some computed or prepared value on to an internal API. We (as devs) are in control of the codepaths and the values, and so the assertion is there to maintain that *our code* satisfies the invariant condition as intended. 

But user-input is user-input. We have no control at all over what users pass in, and can't assume anything about it. So in general, it's not appropriate to make assertions on values that come directly from users (`out.shape` in this case, since `out` come as-is from a parameter of a public API function). Instead, we can perform input validation, but validation of user-input should raise descriptive exceptions, to tell the user what *they* did that was wrong or unexpected.

Reply:
This is a very weird one, the same exact thing doesn't work when a is float!
`>>> a = np.array([1,2,3])
>>> np.cumsum(a,out=b)
array([[1., 0., 0.],
       [3., 0., 0.],
       [6., 0., 0.]])
>>> a = np.array([1,2,3])
>>> b = np.zeros((3,3))
>>> np.cumsum(a,out=b)
array([[1., 0., 0.],
       [3., 0., 0.],
       [6., 0., 0.]])
>>> a = np.array([1.1,2.2,3.3])
>>> b = np.zeros((3,3))
>>> np.cumsum(a,out=b)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in cumsum
  File "/rkarimi_l/miniconda2/envs/legate/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 2569, in cumsum
    return _wrapfunc(a, 'cumsum', axis=axis, dtype=dtype, out=out)
  File "/rkarimi_l/miniconda2/envs/legate/lib/python3.8/site-packages/numpy/core/fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
ValueError: provided out is the wrong size for the accumulation.
`
As for the type of error handling, yes, it should not have been an assert since not an internal error. Although considering the unusual behavior of numpy I'm not sure if I should set it as `NotImplementedError()` or not. For now I'll be setting it as `RuntimeError()`.
e01c5b9cb25fe48ddd63d0e6869b45c44c50133e